### PR TITLE
Adding groupnorm TPP

### PR DESCRIPTION
**Pull request by kvoronin-intel**
Summary:

Adding groupnorm TPP (refactored equation/equatiuon_groupnorm.c sample) similar to what was done for batchnorm earlier.

### CODE REVIEW

File: samples/deeplearning/op_lib_tpp/groupnorm_tpp.h [hunk 1]
```diff
@@ -0,0 +1,1419 @@
...
+
+  libxsmm_blasint ldo = bc;
+  libxsmm_blasint ld  = bc;
+  libxsmm_blasint tmp_ld, tmp_ld2;
+  libxsmm_blasint my_eqn10;
+
+  libxsmm_meltw_unary_shape  unary_shape;
+  libxsmm_meltw_binary_shape binary_shape;
+
+  libxsmm_meltw_unary_flags   unary_flags;
```
Comment (alheinecke) [inline samples/deeplearning/op_lib_tpp/groupnorm_tpp.h:100]: These variables should be of type libxsmm_bitfield

File: samples/deeplearning/op_lib_tpp/groupnorm_tpp.h [hunk 2]
```diff
@@ -0,0 +1,1419 @@
...
+  res.num_HW_blocks = (res.H > res.W ? res.H : res.W );
+  res.threads       = threads;
+  res.fuse_type     = fuse_type;
+
+  /* setting up the barrier */
+  res.barrier = libxsmm_barrier_create(threads, 1);
+
+  /* TPP creation */
+  ldo = res.G;
+  unary_shape.m           = res.G;
```
Comment (alheinecke) [inline samples/deeplearning/op_lib_tpp/groupnorm_tpp.h:135]: here you can use the newly added shape creation functions (if you like):
https://github.com/libxsmm/libxsmm/blob/master/include/libxsmm.h#L552

**Review by alheinecke — COMMENTED** [review]
LGTM:
I think the latest merge from master yesterday brought these changes that flags are now of type libxsmm_bitfield

File: samples/deeplearning/op_lib_tpp/groupnorm_tpp.h (hunk 1 above)
Comment (kvoronin-intel) [reply samples/deeplearning/op_lib_tpp/groupnorm_tpp.h]: Done.

File: samples/deeplearning/op_lib_tpp/groupnorm_tpp.h (hunk 2 above)
Comment (kvoronin-intel) [reply samples/deeplearning/op_lib_tpp/groupnorm_tpp.h]: Yep, I guess I looked previously at some old samples where these shape creation APIs were not used. Now it's fixed.

**Review by alheinecke — APPROVED** [review]
I'm fine with merging, just holding off for now as it's marked WIP

File: samples/deeplearning/op_lib_tpp/groupnorm_tpp.h [hunk 3]
```diff
@@ -0,0 +1,1354 @@
...
+        eqn_param.output.primary = &LIBXSMM_VLA_ACCESS(4, din, np, cp, hwb*(HW/num_HW_blocks), 0, CP, HW, CB);
+        cfg.din_func(&eqn_param);
+      }
+    }
+
+    libxsmm_barrier_wait(cfg.barrier, ltid); /* not needed? */
+
+    for ( cp = thr_begin_C; cp < thr_end_C; ++cp ) {
+      all_zero_param.out.primary = &LIBXSMM_VLA_ACCESS(2, dgamma, cp, 0, CB);
+      cfg.all_zero_kernel(&all_zero_param);
```
Comment (kvoronin-intel) [inline samples/deeplearning/op_lib_tpp/groupnorm_tpp.h:1214]: Missing part in the older sample equation/equation_groupnorm.cpp, which (after moving to new tpp code) led to incorrect computations for Resnet-50.

**Comment by kvoronin-intel** [comment]
> I'm fine with merging, just holding off for now as it's marked WIP

After fixing a couple of issues (one major for the missing initialization of certain arrays in bwd), the PR is ready to get merged in my opinion.

**Review by alheinecke — APPROVED** [review]

**Pull request merged**